numpy/core/src/umath/logical_gufuncs.c.src

```diff
@@ -0,0 +1,607 @@
+        a = *(TYPE *)a_i;                       \
+        b = *(TYPE *)b_i;                       \
+        if (OP(a,b)){                           \
+            a_i += a_I;                         \
+            b_i += b_I;                         \
+        } else {                                \
+            *((npy_bool *)c_n) = NPY_FALSE;     \
+            break;                              \
+        }                                       \
+    }
+
+#define any_INNER_LOOP(OP, TYPE)                \
+    *((npy_bool *)c_n) = NPY_FALSE;             \
+    for (i = 0; i < I; i++){                    \
+        a = *(TYPE *)a_i;                       \
+        b = *(TYPE *)b_i;                       \
+        if (OP(a,b)){                           \
+            *((npy_bool *)c_n) = NPY_TRUE;      \
+            break;                              \
+        } else {                                \
+            a_i += a_I;                         \
+            b_i += b_I;                         \
+        }                                       \
+    }
+
+/* -------------------------------------------------------------------------- */
+/* create the family of functions using a template  */
+
+/**begin repeat
+* #TYPE = byte,ubyte,short,ushort,int,uint,long,ulong,longlong,ulonglong,float,
+*         double,longdouble,cfloat,cdouble,clongdouble#
+* #OP_PREFIX = R,R,R,R,R,R,R,R,R,R,R,R,R,PyArray_C,PyArray_C,PyArray_C#
+*/
+
+/**begin repeat1
+* #OP = EQ,NE,LT,LE,GT,GE#
+* #OPNAME = equal,not_equal,less,less_equal,greater,greater_equal#
+*/
+
+/**begin repeat2
+* #ALL_OR_ANY = all,any#
+*/
+
+static void base_@TYPE@_@ALL_OR_ANY@_@OPNAME@
+        (char **args, npy_intp *dimensions,
+         npy_intp* steps, void* data)
+{
+    npy_intp i, n;
+    npy_intp N = dimensions[0], I = dimensions[1];
+    char *a_n = args[0], *b_n = args[1], *c_n = args[2];
+    npy_intp a_N = steps[0], b_N = steps[1], c_N = steps[2];
+    npy_intp a_I = steps[3], b_I = steps[4];
+
+    char *a_i, *b_i;
+    npy_@TYPE@ a, b;
+
+    for (n = 0; n < N; n++) {
+        a_i = a_n;
+        b_i = b_n;
+
+        @ALL_OR_ANY@_INNER_LOOP(@OP_PREFIX@@OP@, npy_@TYPE@)
+
+        a_n += a_N;
+        b_n += b_N;
+        c_n += c_N;
+    }
+}
+
+/**end repeat2**/
+/**end repeat1**/
+/**end repeat**/
+
+
+/* -------------------------------------------------------------------------- */
+/* vectorized versions*/
+
+#define BLOCK_SIZE 1024
+#define ALIGNMENT_SIZE 32
+#define ALIGNED(p) ((unsigned long)p & ALIGNMENT_SIZE)
+
+/*
+*  The core array must be contiguous, ie the core dimension steps must
+*  be equal to the sizeof the type
+*
+*  TODO: all contiguous signed and unsigned integer types could be handled by
+*  a single vectorized all_equal function.  Besides there should be at least 
+*  one simple C function without src templating.
+*/
+/**begin repeat
+* #TYPE = float,double,int#
+**/
+void simd_core_contiguous_@TYPE@_all_equal(
```

Comment:
this needs the `NPY_GCC_OPT_3` attribute to actually vectorize with GCC

    void NPY_GCC_OPT_3 simd_core_contiguous_@TYPE@_all_equal(